text/0000-builder-spec.md

```diff
@@ -0,0 +1,270 @@
+
+A sample inspected builder (using `pack inspect-builder`), looks like:
+```bash
+ $  pack inspect-builder cnbs/sample-builder:bionic
+Inspecting builder: cnbs/sample-builder:bionic
+
+REMOTE:
+
+Created By:
+  Name: Pack CLI
+  Version: 0.13.1+git-4134cc6.build-1135
+
+Trusted: No
+
+Stack:
+  ID: io.buildpacks.samples.stacks.bionic
+
+Lifecycle:
+  Version: 0.9.1
+  Buildpack APIs:
+    Deprecated: (none)
+    Supported: 0.2, 0.3, 0.4
+  Platform APIs:
+    Deprecated: (none)
+    Supported: 0.3, 0.4
+
+Run Images:
+  cnbs/sample-stack-run:bionic
+
+Buildpacks:
+  ID                            VERSION        HOMEPAGE
+  samples/java-maven            0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/java-maven
+  samples/kotlin-gradle         0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/kotlin-gradle
+  samples/ruby-bundler          0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/ruby-bundler
+  samples/hello-universe        0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-universe
+  samples/hello-world           0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-world
+  samples/hello-moon            0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-moon
+
+Detection Order:
+ ├ Group #1:
+ │  └ samples/java-maven@0.0.1
+ ├ Group #2:
+ │  └ samples/kotlin-gradle@0.0.1
+ ├ Group #3:
+ │  └ samples/ruby-bundler@0.0.1
+ └ Group #4:
+    └ samples/hello-universe@0.0.1
+       └ Group #1:
+          ├ samples/hello-world@0.0.1
+          └ samples/hello-moon@0.0.1
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+## Structure of Builders
+### Files/Directories
+A builder is composed of at least the following directories/files:
+```
+- /cnb/buildpacks/...<buildpack ID>/<buildpack version>/
+- /cnb/lifecycle/<lifecycle binaries>
+- /cnb/order.toml
+- /cnb/stack.toml
+- /layers
+- /platform/env/<env-files>
+- /workspace
+```
+
+### Env Vars/Labels
+A builder's environment is the build-time environment of the stack, and as such, requires all of the [build image specifications][build-image-specs], including:
+* The image config's User field is set to a non-root user with a writable home directory.
+* The image config's Env field has the environment variable CNB_STACK_ID set to the stack ID.
+* The image config's Env field has the environment variable CNB_USER_ID set to the user †UID/‡SID of the user specified in the User field.
+* The image config's Env field has the environment variable CNB_GROUP_ID set to the primary group †GID/‡SID of the user specified in the User field.
+* The image config's Label field has the label io.buildpacks.stack.id set to the stack ID. (string)
+* The image config's Label field has the label io.buildpacks.stack.mixins set to a JSON array containing mixin names for each mixin applied to the image.
```

Comment:
Instead of enumerating the requirements of a `build-image` maybe it is best to simply leave the link to the spec, so that this section doesn't fall out of date.

Reply:
41b9a4ae471e3490de686f66e81fa8bbd999a262